### PR TITLE
Don't copy data or use standalone objects during validation and object creation

### PR DESCRIPTION
**Pull request by alazier**
These changes are needed for the JS binding as it requires support for object creation from containers when using the dynamic interface.

We now no longer used standalone objects when creating persisted objects. We do validation of each property as objects are being created which means we no longer need to do an extra allocation and initialization of standalone objects when creating nested objects. So this should speed things up in theory.

A number of small bugs were fixed and the tests were updated accordingly. There was one issue where objects would be updated even without using one of the update apis. update and standalone promotion flags are now propagated to all nested objects even through objects without a primary key.

### CODE REVIEW

**Comment by bdash** [comment]
IMO it'd be preferable to keep using an enum rather than a bool, even though you've eliminated most of the members. `RLMCreationOptionsUpdateOrCreate` is much more descriptive at a call site than `true`.

**Comment by alazier** [comment]
I agree having `RLMCreationOptions` is better than a bool. My main issue with this was that the options we had were confusing and were set and mutated in odd ways - removing them made cleaning up the code a lot easier. I'm happy to add this back but we will only support two options `RLMCreationOptionsUpdateOrCreate` and `RLMCreationOptionsPromoteStandalone`.

**Comment by segiddins** [comment]
I personally don't like referring to `literals`, since all we care about those objects is that they're KVC-compliant

**Comment by alazier** [comment]
Not quite true. We also support array literals.

**Comment by bdash** [comment]
Literals doesn't seem like the appropriate term. At the point we're dealing with the objects its irrelevant whether they were literals or created in some other fashion.

**Comment by segiddins** [comment]
Since Jenkins doesn't seem to have run on this, I'm also concerned that this might break our recent improvements with object re-use.

**Comment by alazier** [comment]
All the tests pass. I'm not sure what the goal of the object re-use pr was but we should be moving away from creating standalone objects internally, except in the case where we are actually initializing standalone objects. This pr accomplishes that in all cases where we are creating persisted objects.

Feel free to suggest alternate terminology. I can't think of anything better offhand.

My main goal here is unblock the JS work. The current functionality in this pr does this and also fixes a few other bugs with the current code. Don't think I am going to spend much more time on this so it would be helpful to figure out what else needs to be done to get this merged.

File: Realm/RLMAccessor.mm [hunk 1]
```diff
@@ -179,25 +179,34 @@ static inline void RLMSetValue(__unsafe_unretained RLMObjectBase *const obj, NSU
     }
 }
 
-static inline size_t RLMAddLinkedObject(RLMObjectBase *link,
-                                        __unsafe_unretained RLMRealm *const realm,
-                                        RLMCreationOptions options) {
+static inline RLMObjectBase *RLMAddLinkedObjectOrLiteral(__unsafe_unretained RLMRealm *const realm,
+                                                         NSString *className,
```
Comment (segiddins) [inline Realm/RLMAccessor.mm]: pretty sure this can be made `__unsafe_unretained`

File: Realm/RLMObjectBase.mm [hunk 2]
```diff
@@ -58,6 +59,73 @@ - (instancetype)init {
...
+            RLMObjectSchema *objSchema = schema[prop.objectClassName];
+            RLMArray *objects = [[RLMArray alloc] initWithObjectClassName: objSchema.className standalone:YES];
+            for (id el in obj) {
+                [objects addObject:RLMValidatedRealmObject(el, schema, objSchema)];
+            }
+            return objects;
+        }
+
+        // if not a literal throw
+        NSString *message = [NSString stringWithFormat:@"Invalid value '%@' for property '%@'", obj ?: @"nil", prop.name];
```
Comment (segiddins) [inline Realm/RLMObjectBase.mm]: no need for the `?:` here

**Comment by tgoyne** [comment]
"Literal" is dumb and wrong but in lieu of a similarly concise term I'd prefer to keep using it over "KVC compliant object or array with values in the correct order".

File: Realm/RLMAccessor.mm [hunk 3]
```diff
@@ -179,25 +179,34 @@ static inline void RLMSetValue(__unsafe_unretained RLMObjectBase *const obj, NSU
...
+    if (link->_realm == realm) {
+        return link;
+    }
+
+    if (creationOptions & RLMCreationOptionsPromoteStandalone) {
+        if (!link->_realm) {
+            RLMAddObjectToRealm(link, realm, creationOptions & RLMCreationOptionsCreateOrUpdate);
+            return link;
         }
+        @throw RLMException(@"Can not add objects from a differnt Realm");
```
Comment (segiddins) [inline Realm/RLMAccessor.mm]: `different`

File: Realm/RLMObjectStore.mm [hunk 4]
```diff
@@ -463,8 +468,91 @@ void RLMAddObjectToRealm(RLMObjectBase *object, RLMRealm *realm, RLMCreationOpti
     RLMInitializeSwiftListAccessor(object);
 }
 
-RLMObjectBase *RLMCreateObjectInRealmWithValue(RLMRealm *realm, NSString *className, id value, RLMCreationOptions options) {
-    if (options & RLMCreationOptionsUpdateOrCreate && RLMIsObjectSubclass([value class])) {
+static void RLMValidateNestedObject(id obj, NSString *className, RLMSchema *schema, bool validateNested, bool allowMissing) {
+    if (obj != nil && obj != NSNull.null) {
+        if (RLMObjectBase *objBase = RLMDynamicCast<RLMObjectBase>(obj)) {
+            RLMObjectSchema *objectSchema = objBase->_objectSchema;
```
Comment (segiddins) [inline Realm/RLMObjectStore.mm]: `__unsafe_unretained const`?

File: Realm/RLMObjectStore.mm [hunk 5]
```diff
@@ -463,8 +468,91 @@ void RLMAddObjectToRealm(RLMObjectBase *object, RLMRealm *realm, RLMCreationOpti
...
+            break;
+        case RLMPropertyTypeObject:
+            if (validateNested) {
+                RLMValidateNestedObject(obj, prop.objectClassName, schema, validateNested, allowMissing);
+            }
+            break;
+        case RLMPropertyTypeArray: {
+            if (validateNested) {
+                if (obj != nil && obj != NSNull.null) {
+                    id<NSFastEnumeration> array = obj;
```
Comment (segiddins) [inline Realm/RLMObjectStore.mm]: this should be a dynamic cast and throw an exception if it fails

File: Realm/Tests/ObjectTests.m [hunk 6]
```diff
@@ -205,12 +205,20 @@ + (NSString *)primaryKey {
 }
 @end
 
+@interface EmployeeWrapper : RLMObject
```
Comment (segiddins) [inline Realm/Tests/ObjectTests.m]: `LinkToPrimaryEmployeeObject`?

File: Realm/RLMObjectStore.mm (hunk 4 above)
Comment (alazier) [inline Realm/RLMObjectStore.mm]: I don't think it is necessary to add this everywhere. It is an optimization and should be used in code paths which need optimization.

File: Realm/RLMObjectStore.mm (hunk 5 above)
Comment (alazier) [inline Realm/RLMObjectStore.mm]: dynamic cast doesn't work with protocols. Can validate that this is enumerable though.

File: Realm/RLMObjectBase.mm [hunk 7]
```diff
@@ -58,26 +59,56 @@ - (instancetype)init {
     return self;
 }
 
+static id RLMValidatedObjectForProperty(id obj, RLMProperty *prop, RLMSchema *schema) {
+    if (!RLMIsObjectValidForProperty(obj, prop)) {
```
Comment (jpsim) [inline Realm/RLMObjectBase.mm]: I'd prefer an early return here.

File: Realm/RLMAccessor.mm [hunk 8]
```diff
@@ -179,25 +179,34 @@ static inline void RLMSetValue(__unsafe_unretained RLMObjectBase *const obj, NSU
...
 }
 
-static inline size_t RLMAddLinkedObject(RLMObjectBase *link,
-                                        __unsafe_unretained RLMRealm *const realm,
-                                        RLMCreationOptions options) {
+static inline RLMObjectBase *RLMGetLinkedObjectForValue(__unsafe_unretained RLMRealm *const realm,
+                                                        __unsafe_unretained NSString *const className,
+                                                        __unsafe_unretained id const value,
+                                                        RLMCreationOptions creationOptions) {
+    RLMObjectBase *link = RLMDynamicCast<RLMObjectBase>(value);
```
Comment (segiddins) [inline Realm/RLMAccessor.mm]: this should also be `__unsafe_unretained const`?

File: Realm/RLMAccessor.mm [hunk 9]
```diff
@@ -644,12 +650,11 @@ void RLMDynamicValidatedSet(RLMObjectBase *obj, NSString *propName, id val) {
...
     }
-    RLMDynamicSet(obj, prop, val, prop.isPrimary ? RLMCreationOptionsEnforceUnique : 0);
+    RLMDynamicSet(obj, prop, val, RLMCreationOptionsPromoteStandalone);
 }
 
-void RLMDynamicSet(__unsafe_unretained RLMObjectBase *const obj,
-                   __unsafe_unretained RLMProperty *const prop,
-                   __unsafe_unretained id val, RLMCreationOptions options) {
+void RLMDynamicSet(__unsafe_unretained RLMObjectBase *const obj, __unsafe_unretained RLMProperty *const prop,
+                   __unsafe_unretained id val, RLMCreationOptions creationOptions) {
```
Comment (segiddins) [inline Realm/RLMAccessor.mm]: `val` needs to be `const`

File: Realm/RLMObjectBase.mm [hunk 10]
```diff
@@ -58,26 +59,56 @@ - (instancetype)init {
...
+        if (array.count != properties.count) {
+            @throw RLMException(@"Invalid array input. Number of array elements does not match number of properties.");
+        }
         for (NSUInteger i = 0; i < array.count; i++) {
-            [self setValue:array[i] forKeyPath:[properties[i] name]];
+            [self setValue:RLMValidatedObjectForProperty(array[i], properties[i], schema) forKeyPath:[properties[i] name]];
         }
     }
     else {
         // assume our object is an NSDictionary or a an object with kvc properties
```
Comment (jpsim) [inline Realm/RLMObjectBase.mm]: "a an"

File: Realm/RLMObjectStore.mm [hunk 11]
```diff
@@ -435,10 +432,13 @@ void RLMAddObjectToRealm(__unsafe_unretained RLMObjectBase *const object,
 
     // get or create row
     bool created;
-    auto primaryGetter = [=](__unsafe_unretained RLMProperty *const p) {
-        return [object valueForKey:p.getterName];
-    };
-    object->_row = (*schema.table)[RLMCreateOrGetRowForObject(schema, primaryGetter, options, created)];
+    auto primaryGetter = [=](RLMProperty *p) { return [object valueForKey:p.getterName]; };
```
Comment (jpsim) [inline Realm/RLMObjectStore.mm]: what required this change to remove the `__unsafe_unretained`?

File: Realm/RLMAccessor.mm (hunk 8 above)
Comment (tgoyne) [inline Realm/RLMAccessor.mm]: Local variables generally don't get pointless retain/releases inserted, so only if profiling indicates a problem.

File: Realm/RLMObjectStore.mm (hunk 11 above)
Comment (alazier) [inline Realm/RLMObjectStore.mm]: rebase error

File: Realm/RLMObjectBase.mm [hunk 12]
```diff
@@ -58,26 +59,56 @@ - (instancetype)init {
...
-        NSDictionary *dict = RLMValidatedDictionaryForObjectSchema(value, _objectSchema, schema);
-        for (NSString *name in dict) {
-            id val = dict[name];
-            // strip out NSNull before passing values to standalone setters
-            if (val == NSNull.null) {
-                val = nil;
+        NSDictionary *defaultValues = nil;
```
Comment (jpsim) [inline Realm/RLMObjectBase.mm]: This code is quite similar to what's in `RLMValidateValueForObjectSchema`, could you write a function?

**Comment by jpsim** [comment]
Please add a CHANGELOG entry for the update issue you've fixed.

File: Realm/RLMAccessor.mm (hunk 8 above)
Comment (alazier) [inline Realm/RLMAccessor.mm]: In general I think we should only be adding optimizations where the profiler indicates there is an issue...

File: Realm/RLMObjectBase.mm (hunk 12 above)
Comment (alazier) [inline Realm/RLMObjectBase.mm]: Don't think it is worth it in this case. The contents of the loop are slightly different so we would either need a block/lambda, or and out paramater for the cached defaults which is awkward.

**Comment by alazier** [comment]
anything else?

File: Realm/RLMObjectStore.mm (hunk 11 above)
Comment (jpsim) [inline Realm/RLMObjectStore.mm]: ok, then please revert this change.

**Comment by jpsim** [comment]
Other than the remaining rebase error, you get my :+1:

File: Realm/RLMAccessor.mm [hunk 13]
```diff
@@ -179,25 +179,34 @@ static inline void RLMSetValue(__unsafe_unretained RLMObjectBase *const obj, NSU
     }
 }
 
-static inline size_t RLMAddLinkedObject(RLMObjectBase *link,
-                                        __unsafe_unretained RLMRealm *const realm,
-                                        RLMCreationOptions options) {
+static inline RLMObjectBase *RLMGetLinkedObjectForValue(__unsafe_unretained RLMRealm *const realm,
```
Comment (tgoyne) [inline Realm/RLMAccessor.mm]: This probably needs to be marked as `NS_RETURNS_RETAINED` to avoid autoreleasing everything (`RLMAddLinkedObject` avoided the autorelease by returning the row index rather than the object).

Comment (tgoyne) [inline Realm/RLMAccessor.mm]: i.e. add `static inline RLMObjectBase *RLMGetLinkedObjectForValue(RLMRealm *realm, NSString *className, id value, RLMCreationOptions creationOptions) NS_RETURNS_RETAINED;` on the line before (GCC doesn't support marking the definition directly so clang gives a warning).

**Comment by segiddins** [comment]
Seems good :+1:

File: Realm/RLMAccessor.mm (hunk 13 above)
Comment (alazier) [inline Realm/RLMAccessor.mm]: Does it matter since it's inlined?

Comment (tgoyne) [inline Realm/RLMAccessor.mm]: Marking it as inline doesn't guarantee that it actually will be inlined.

File: Realm/RLMObjectStore.mm (hunk 11 above)
Comment (alazier) [inline Realm/RLMObjectStore.mm]: I made this change in the other spot RLMCreateOrGetRowForObject was called...

Will make a new pr with this fix.
